cvmfs/network/download.cc

```diff
@@ -1888,15 +1916,68 @@ Failures DownloadManager::Fetch(JobInfo *info) {
     //          info->wait_at[0], info->wait_at[1]);
     pipe_jobs_->Write<JobInfo*>(info);
 
+    bool is_running = true;
     do {
       DataTubeElement* ele = info->GetDataTubePtr()->PopFront();
 
-      if (ele->action == kActionStop) {
-        delete ele;
+      switch (ele->action) {
+        case kActionStop:
+          is_running = false;
+        break;
+        case kActionEndOfData:
+          /*
+           * End of data processing.
+           * Used for synchronization before calling VerifyAndFinalize()
+           * Next element action received should be kActionStop
+           */
```

Comment:
If we do nothing, why does this command need to be sent?

Reply:
because the "gate"/synchronization of when to run `VerifyAndFinalize` is based on the empty data tube. 
but you will remove an element before decompressing it. during decompression errors can occur.  as such we need one last item to tell us "we are done with all data processing"